adblock-lean

```diff
@@ -306,72 +311,77 @@ generate_preprocessed_blocklist_file_parts()
-
-					preprocessed_blocklist_line_count=$(( preprocessed_blocklist_line_count + blocklist_file_part_line_count ))
-					blocklist_id=$((blocklist_id+1))
-					continue 2
-				else
-					log_msg "Downloaded blocklist file part line count: ${blocklist_file_part_line_count} less than configured minimum: ${min_blocklist_file_part_line_count}."
-					break
-				fi
+			if [[ "${rogue_element_action}" != "IGNORE" ]]
+			then
+				log_msg "Checking for any rogue elements."
+				tee >(grep -Evn -m1 '^(local=/[[:alnum:]*][[:alnum:]*_.-]+/|bogus-nxdomain=[0-9.]+){0,1}$' > /tmp/rogue_element)
 			else
+				cat
+			fi | gzip > /tmp/blocklist.${blocklist_id}.gz
+
+			blocklist_file_part_size_KB=$(( ($(cat /tmp/adblock-lean_part_size_B 2>/dev/null) + 0) / 1024 ))
+			blocklist_file_part_line_count="$(cat /tmp/adblock-lean_part_line_count)"
+			: "${blocklist_file_part_line_count:=0}"
+
+			rm -f /tmp/adblock-lean_part_size_B /tmp/adblock-lean_part_line_count 2>/dev/null
+
+			if ! grep -q "Download completed" /tmp/uclient-fetch_err
+			then
 				log_msg "Download of new blocklist file part from: ${blocklist_url} failed."
-				if [[ -f "/tmp/blocklist.${blocklist_id}" ]]
+				if [[ "${blocklist_file_part_size_KB}" -ge "${max_blocklist_file_part_size_KB}" ]]
 				then
-					blocklist_file_part_size_KB=$(get_file_size_KB /tmp/blocklist.${blocklist_id})
-					if [[ "${blocklist_file_part_size_KB}" -eq "${max_blocklist_file_part_size_KB}" ]]
-					then
 						log_msg "Downloaded blocklist file part size exceeded the maximum value set in config (${max_blocklist_file_part_size_KB} KB)."
 						log_msg "Consider either increasing this value in the config or removing the correasponding blocklist url."
 						break
-					fi
 				fi
+				sleep_before_retry
+				continue
 			fi
-			log_msg "Sleeping for 5 seconds after failed download attempt."
-			sleep 5
+
+			if read -r rogue_element < /tmp/rogue_element
+			then
+				log_msg "Rogue element: '${rogue_element}' identified originating in blocklist file part from: ${blocklist_url}."
+
+				if [[ "${rogue_element_action}" == "STOP" ]]
+				then
+					return 1
+				else
+					log_msg "Skipping file part and continuing."
+					rm -f /tmp/rogue_element
+					continue 2
+				fi
+			fi
+
+			if [[ "${blocklist_file_part_line_count}" -ge "${min_blocklist_file_part_line_count}" ]]
+			then
+				log_msg "Download of new blocklist file part from: ${blocklist_url} succeeded (downloaded file size: ${blocklist_file_part_size_KB} KB; line count: ${blocklist_file_part_line_count})."
+
+				log_msg "Sanitizing blocklist file part."
+
+				rm -f /tmp/rogue_element
+
+				preprocessed_blocklist_line_count=$(( preprocessed_blocklist_line_count + blocklist_file_part_line_count ))
+				blocklist_id=$((blocklist_id+1))
+				continue 2
+			else
+				log_msg "Downloaded blocklist file part line count: ${blocklist_file_part_line_count} less than configured minimum: ${min_blocklist_file_part_line_count}."
+				break
+			fi
+			sleep_before_retry
```

Comment:
Let’s just expand this out without function.

Reply:
> Let’s just expand this out without function

The code in that function (sleep_before_retry) is called twice in this PR. Currently only the first call is strictly necessary because if everything works as intended, the second call will never get executed. It is in place because otherwise, if some of the logic is changed at a later point and execution is allowed to reach the end of the while loop, you will have a bug where download is retried without a delay.

So in this instance, the function is there to reduce code duplication. Do you still want to replace calls to it with duplicate code?

Reply:
Yes. We can also from the calls to safeguard against failure in future. I’d rather redesign the logic entirely than build in safeguards against not properly following the logic in the future.

Reply:
Sure, I will change this. However I'd like to give you an example of this safeguard applied in the wild to demonstrate that it's not far-fetched. The example is nftables and iptables chain policies. You can not create a base chain without a default policy. Why - because humans make mistakes, and a good coder recognizes this fact. If a subtle mistake may lead to software going into undetermined state then a good practice is to put safeguards in place to prevent that. I won't insist on this safeguard here but I wanted to explain the logic behind it and the fact that it's standard practice.

Reply:
> Yes. We can also remove the repetition to safeguard against failure in future.

Done.

Reply:
Thanks. Should be able to test this shortly.